### DISCUSSION TITLE
inserting signature using .crt files with x509 encryption

### DISCUSSION THREAD

**Discussion by jumpingllamas** [MatthiasValvekens/pyHanko#190] [Q&A]
hi, does anyone have a sample code on how to insert x509 crypto signature objects into existing signature fields in a pdf file?

**Comment by MatthiasValvekens** [comment]
Hi @jumpingllamas,


The [first example in the docs](https://pyhanko.readthedocs.io/en/latest/lib-guide/signing.html#a-simple-example) fits the bill. If you're confused about the file extensions/terminology, here's what you have to supply:

 - Your own X.509 certificate (that's probably the .crt file you mentioned)
 - The private key corresponding to your certificate
 - Any relevant intermediate CA certificates that might be necessary to validate your certificate (optional, but recommended in >90% of cases)

The certificates will be embedded into the document. The private key will (obviously) only be used to _generate_ the signature, and can also be provided indirectly (e.g. through a hardware token). Nonetheless, an X.509 certificate _alone_ is not enough to produce a signature, you also need the corresponding private key.

(Terminological note: X.509 is the name of the standard used to format certificates, it doesn't specify any particular cryptographic mechanism or even care about private keys at all.)

Hope this helps.

**Reply by jumpingllamas** [reply]
Thanks for the explanation, was trying to wrap my head around how to load and populate it in all the signature fields created.
